IdentityCore/src/requests/MSIDNonceTokenRequest.m

```diff
@@ -0,0 +1,205 @@
+         {
+             
+             if (openIdError)
+             {
+                 completionBlock(nil, openIdError);
+                 return;
+             }
+             
+             [self executeNetworkRequestWithCompletion:completionBlock];
+         }];
+    }];
+}
+
+- (void)executeNetworkRequestWithCompletion:(nonnull MSIDNonceRequestCompletion)completionBlock
+{
+    MSIDHttpRequest *nonceRequest = [self configureNonceNetworkRequestForEndpoint:self.requestParameters.tokenEndpoint context:self.requestParameters];
+    [nonceRequest sendWithBlock:^(NSDictionary *response, NSError *error)
+    {
+        if (error)
+        {
+            if (completionBlock) completionBlock(nil, error);
+            return;
+        }
+        
+        if (![response isKindOfClass:[NSDictionary class]])
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, self.requestParameters, @"Unexpected nonce response received");
+            NSError *nwError = MSIDCreateError(MSIDErrorDomain, MSIDErrorServerInvalidResponse, @"Unexpected nonce response", nil, nil, nil, nil, nil, YES);
+            if (completionBlock) completionBlock(nil, nwError);
+            return;
+        }
+        
+        NSString *nonce = [response msidStringObjectForKey:@"Nonce"];
+        
+        if ([NSString msidIsStringNilOrBlank:nonce])
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, self.requestParameters, @"Didn't receive valid nonce in response");
+            NSError *nwError = MSIDCreateError(MSIDErrorDomain, MSIDErrorServerInvalidResponse, @"Didn't receive valid nonce in response", nil, nil, nil, nil, nil, YES);
+            if (completionBlock) completionBlock(nil, nwError);
+            return;
+        }
+        
+        [self.class cacheNonceForKey:self.requestParameters.authority.environment nonce:nonce];
+        if (completionBlock)
+        {
+            completionBlock(nonce, nil);
+        }
+    }];
+}
+
+- (MSIDHttpRequest *)configureNonceNetworkRequestForEndpoint:(NSURL *)endpoint context:(id<MSIDRequestContext>)context
+{
+    if (!endpoint)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"No endpoint provided to get nonce from!");
+        NSParameterAssert(endpoint);
+        return nil;
+    }
+    
+    MSIDHttpRequest *request = [[MSIDHttpRequest alloc] init];
+    NSMutableURLRequest *urlRequest = [NSMutableURLRequest new];
+    urlRequest.URL = endpoint;
+    urlRequest.HTTPMethod = @"POST";
+    request.urlRequest = urlRequest;
+    
+    __auto_type requestConfigurator = [MSIDAADRequestConfigurator new];
+    [requestConfigurator configure:request];
+    
+    NSMutableDictionary *parameters = [NSMutableDictionary new];
+    
+    parameters[MSID_OAUTH2_GRANT_TYPE] = @"srv_challenge";
+    [parameters addEntriesFromDictionary:parameters];
+    request.parameters = parameters;
+    request.urlRequest = urlRequest;
```

Comment:
`request.urlRequest = urlRequest;` is called above once, should we remove the first one?

Reply:
Yep, thanks. Removed